src/auxiliary/mpi_arrays.jl

```diff
@@ -0,0 +1,180 @@
+# LoopVectorization.jl etc.
+# See https://juliaarrays.github.io/ArrayInterface.jl/stable/
+ArrayInterface.parent_type(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = Parent
+
+
+# TODO: MPI. Do we need LinearAlgebra methods such as `norm` or `dot`?
+
+
+# `mapreduce` functionality from Base using global reductions via MPI communication
+function Base.mapreduce(f::F, op::Op, u::TrixiMPIArray; kwargs...) where {F, Op}
+  local_value = mapreduce(f, op, parent(u); kwargs...)
+  return MPI.Allreduce(local_value, op, mpi_comm(u))
+end
+
+
+# TODO: MPI. Default settings of OrdinaryDiffEq etc.
+# Interesting options could be
+# - ODE_DEFAULT_UNSTABLE_CHECK
+# - ODE_DEFAULT_ISOUTOFDOMAIN
+# - ODE_DEFAULT_NORM
+# See https://github.com/SciML/DiffEqBase.jl/blob/master/src/common_defaults.jl
+
+# Problems and inconsistencies with TrixiMPIArrays
+# A basic question is how to handle `length`. We want `TrixiMPIArray`s to behave
+# like regular `Array`s in most code, e.g., when looping over an array (which
+# should use `eachindex`). At the same time, we want to be able to use adaptive
+# time stepping using error estimates in OrdinaryDiffEq.jl. There, the default
+# norm `ODE_DEFAULT_NORM` is the one described in the book of Hairer & Wanner,
+# i.e., it includes a division by the global `length` of the array. We could
+# specialize `ODE_DEFAULT_NORM` accordingly, but that requires depending on
+# DiffEqBase.jl (instead of SciMLBase.jl). Alternatively, we could implement
+# this via Requires.jl, but that will prevent precompilation and maybe trigger
+# invalidations. Alternatively, could implement our own norm and pass that as
+# `internalnorm`. Here, we decide to use the least intrusive approach for now
+# and specialize `length` to return a global length while making sure that all
+# local behavior is still working as expected (if using `eachindex` instead of
+# `1:length` etc.). This means that `eachindex(u) != Base.OneTo(length(u))` for
+# `u::TrixiMPIArray` in general, even if `u` and its underlying array use
+# one-based indexing.
+Base.eachindex(u::TrixiMPIArray) = eachindex(parent(u))
+
+function Base.length(u::TrixiMPIArray)
+  local_length = length(parent(u))
+  return MPI.Allreduce(local_length, +, mpi_comm(u))
+end
+
+"""
+    local_length(u)
+
+Like `length(u)`, but returns the length of the local data for `u::TrixiMPIArray`.
+"""
+local_length(u) = length(u)
+local_length(u::TrixiMPIArray) = length(parent(u))
+
+
+# There are some issues when calling `similar` or `copy` only on a single rank,
+# e.g., when using AMR. This reduces to constructing a `TrixiMPIArray` only on
+# a single rank, even if all fields are passed explicitly. For example,
+# ```julia
+# julia> using Trixi
+
+# julia> if Trixi.mpi_isroot()
+#            Trixi.TrixiMPIArray{Float64, 1, Vector{Float64}}(zeros(7,), Trixi.MPI.COMM_WORLD, 5)
+#        end
+# ```
+# works on the second rank but not on the first rank when running with
+# `tmpi 2 julia --threads=1`.
```

Comment:
@sloede Do you have any idea what might be going wrong here?

Reply:
Nope, but I also do not fully understand the issue. Why do you set the rank to `5` inside an if that only executes on rank `0`? Sorry, I am very confused :-)

Reply:
These were just some dummy values (that are not used anywhere). The problem is that this code hangs in `tmpi` when using more than one rank - it hangs on the root executing this stuff, independent of the arguments I set there.

Reply:
Thus, it hangs when you execute `copy(u_mpi)` or `similar(u_mpi)`, but only on root, while the others progress? Hm. This does not sound to me like it makes sense, since copying an MPI communicator should not cause any issues. Just to be sure: Have you run something like
```julia
MPI.Barrier(MPI.COMM_WORLD)
println("BEFORE: ", mpi_rank(u_mpi))
c = copy(u_mpi)
MPI.Barrier(MPI.COMM_WORLD)
println("AFTER: ", mpi_rank(u_mpi))
```
to be sure it really hangs where you think it does?

Reply:
Found the issue. Here, returning the values calls `show`, dispatching to `Base.show_vector`, which uses `length`... Our global `length` tries to communicate, which does not work if called on more than one rank... Fixed in 23d45208

Reply:
Great you found it! However, I also count this as one more caveat of using the `length-is-global` approach we currently pursue.

Reply:
Yeah, definitely...